### DISCUSSION TITLE
Native MacOS App

### DISCUSSION THREAD

**Discussion by Sanster** [Sanster/IOPaint#314] [Announcements]
OptiClean is officially released! Early Waitlist participants will progressively get promo codes. As the App Store can't create 50% discount code, you'll receive a complimentary promo code instead.

App Store Link: https://apps.apple.com/cn/app/id6452387177

---------

Hello everyone, I plan to develop a native macOS app that is completely native from UI to model, no JS/Python/PyTorch involved.
In my test, it runs mush faster on the GPU of the M2 chip than using pytorch(mps device).

https://github.com/Sanster/lama-cleaner/assets/3998421/d2c9efc6-b4fa-4469-b9c3-40794e891735

There is still a lot of work to be done before officially releasing this app. If you are interested, you can fill out [this form](https://docs.google.com/forms/d/e/1FAIpQLSd0YQP_xud5tjKwkXScnanQevtxOJKt2hopeIjwMDdzOxOrdA/viewform) and submit your email address. You will receive a reminder when the app is launched and receive a 50% discount code. Thanks!

**Comment by vicento** [comment]
an interesting App

**Comment by mallman** [comment]
Hi @Sanster.

I've also created a Mac (and iPad) app for LaMa. It's based on a port to Core ML I made with a SwiftUI interface. I built this with the intention of demoing it to a team for the company I work for, but that probably won't happen. I don't know. Maybe it will. I have no intention of publishing it. My app also works with Apple's (hidden) implementation of inpainting, as a pluggable alternative. Apple's inpainting filter is much faster, but from my experience it doesn't give good results. LaMa gives sub-second results on 45 MP images with Core ML, which is certainly adequate.

I've put all my work on github in several private repos. I'd be happy to open them to you if you're interested. Given your interest in commercializing your work I can understand why you would not want to see it though. I generally do not publish my work, because I do not want to be bothered with supporting it.

I'd also be interested to know if you solved one particular problem I ran into in my port to Core ML. Although Apple's Core ML converter supports flexible input image sizes, I have not been able to implement that for LaMa. I won't explain, because you probably know what I'm talking about. So my model is based on a fixed image input size.

I suspect this problem can be resolved by augmenting the Pytorch model with a preprocessing step, but I have not bothered trying. It's too much work for me, and frankly it's probably beyond my abilities. You seem more than capable. If you do solve it, I hope I can entice you to share your solution.

One more thing. Lama cleaner itself made my effort to port LaMa to Core ML much much easier. Thank you for sharing this project. If I ever do publish my work I will certainly credit Lama cleaner where credit is due.

**Reply by Sanster** [reply]
You're right, the LaMa model can only be converted to coreml with a fixed inference size. Before performing model inference, I do some preprocessing. You can refer to the preprocessing method of the [fcf model](https://github.com/Sanster/lama-cleaner/blob/5d988dc5fd1aadff6166a9b6d2101e19b1f62a48/lama_cleaner/model/fcf.py#L1664), which also has a fixed input size. Perhaps there are better methods, but this is how I currently implemented it.

**Reply by mallman** [reply]
I decided to publish my model conversion script. It is very simple. I hope it can be a place to consolidate community efforts to create a high-quality port of LaMa to Core ML. I hope we can get flexible input sizes supported. I will give that another go. The project deps are a little outdated. For example, it uses the coremltools 6.3, and I see there's a version 7 now. Maybe that will make it easier to implement the flexible input sizes.

Anyway, my repo can be found at https://github.com/mallman/CoreMLaMa. I welcome your feedback and contributions.

**Reply by czkoko** [reply]
This project has achieved a flexible shape of stable diffusion, which may be available for reference.

https://github.com/GuernikaCore/GuernikaModelConverter

**Comment by Sanster** [comment]
- web https://opticlean.io/
- TestFlight beta test(Apple Silicon only) link: https://testflight.apple.com/join/B1NNLr92

**Reply by wenber-yu** [reply]
Installazione TestFlight fallita
<img width="359" alt="QQ20230817-112840@2x" src="https://github.com/Sanster/lama-cleaner/assets/7148309/003c7dec-9eb8-4b0c-a47f-c3bbb5c62758">

**Reply by Sanster** [reply]
I have successfully installed it in my environment, and I also see that some users have successfully installed it in the TestFlight backend. I guess it's a network issue, or a problem with the TestFlight server 🤔.

**Reply by mallman** [reply]
With respect to the model encryption issue, can you deploy an unencrypted model?

**Reply by Sanster** [reply]
I am planning to do so, and I have found another related [discussion](https://developer.apple.com/forums/thread/716989). It seems that Apple's encryption server is not very stable.

**Reply by obeone** [reply]
Hi,
What's are specs required ? Does it needs a Apple M* cpu ? Because Testflight refuse the installation ("OptiClean is not compatible with this device").

(My configuration is a Macbook pro 2020. But maybe it's because I'm on Sonoma beta)

**Reply by Sanster** [reply]
Thank you for your interest, but unfortunately, this application is Apple Silicon only.

**Comment by zhi8617** [comment]
請問之後會有圖片縮放跟上傳mask 的功能嗎 ??

**Reply by Sanster** [reply]
图片缩放/平移在计划中，上传 mask 没想好要不要做，感觉需求不是很大 🤔

**Reply by zhi8617** [reply]
謝謝您的回復 , 目前使用起來很好 。  MASK 對我來是是需要的 ，不過您是開發者難度還是您自己知道。 

另外我想在請問您上方的影片示範是真實速度嗎 ? 

我自己是用m1 會要個一兩秒 , 即使是小圖片 。
也想知道在普通pc上面的速度。有在考慮去組一台pc來跑看看 ，不知道會不會差距很多

**Reply by Sanster** [reply]
是真实速度，视频我是用 M2 MAX 录制的，所以可能快一些，手头没有 M1 测试，感谢你的反馈。PC 的话可以直接安装这个开源项目 Lama Cleaner，CPU 推理估计两三秒，GPU 的话小几百毫秒。上传 Mask 的功能 Lama Cleaner 也是有的，mac 上也能安装 `pip install lama-cleaner`：https://github.com/Sanster/lama-cleaner/releases/tag/0.26.0

**Comment by wenber-yu** [comment]
收到上线邮件了, 使用兑换码成功了, 但是下载app后, 没有效果, 是还需要购买对吗?

**Reply by Sanster** [reply]
不需要再购买的。「没有效果」是指还是停留在 Free Trial 页面吗？有可能是网络问题，没检查到激活状态，cmd+q 退出以后再试试

**Reply by Sanster** [reply]
抱歉，我又检查了一下，不是网络的问题。是昨晚发送的第二批 Promo Code 我未进行正确的 products 关联，刚才检查了一下已经关联好了。你不需要做任何其它操作。如果还跳出来付费界面，可能是查询付费状态的本地缓存，我不确定这个会缓存多久。。。如果有其他更新我会反馈在这里。

----English Version------
If you redeemed the Promo Code but the payment page still pops up, it is due to my operational error: Last night I didn't correctly associate the second batch of Promo Codes with the products, which has been fixed today. You don't need to do anything else. If the payment page still pops out, it could be due to the local cache of the querying payment status, but I'm not sure how long it will cache... If there are any other updates, I will post them here.

**Reply by wenber-yu** [reply]
卸载重新安装后, 点击了恢复购买可以了. 感谢.

**Comment by ljdang** [comment]
Android端是否有可行方案呢，有没有计划上线

**Reply by Sanster** [reply]
没有移动端的开发计划

**Comment by wenber-yu** [comment]
是不是可以再做一个扩展图像的？感觉也很常用。

**Reply by Sanster** [reply]
图像扩展能力需要 Stable Diffusion，OptiClean 不会添加，Lama Cleaner 会在下个版本加上 outpainting 功能

**Reply by wenber-yu** [reply]
嗯，我的意思是您可以再单独做一个专用于扩展图像的客户端，就像OptiClean一样，这块使用人群应该也是很大的。
